File: lexer/lexer.go
package lexer

import (
	"gorb/token"
	"strings"
)

func New(input string) *Lexer {
	l := &Lexer{input: input, pos: token.Pos{Ln: 1, Col: 0}}
	l.readChar()
	return l
}

type Lexer struct {
	input        string
	position     int
	readPosition int
	ch           byte
	pos          token.Pos
}

func (l *Lexer) readChar() {
	if l.readPosition >= len(l.input) {
		l.ch = 0
	} else {
		l.ch = l.input[l.readPosition]
	}
	l.position = l.readPosition
	l.readPosition++
	l.pos.Col++
}

func (l *Lexer) NextToken() token.Token {
	var tok token.Token

	l.skipWhitespace()

	switch l.ch {
	case '=':
		if l.peekChar() == '=' {
			ch := l.ch
			l.readChar()
			literal := string(ch) + string(l.ch)
			tok = token.Token{Type: token.EQUALS, Literal: literal}
		} else {
			tok = l.newToken(token.ASSIGN)
		}
	case '+':
		tok = l.newToken(token.PLUS)
	case '-':
		tok = l.newToken(token.MINUS)
	case '!':
		if l.peekChar() == '=' {
			ch := l.ch
			l.readChar()
			literal := string(ch) + string(l.ch)
			tok = token.Token{Type: token.NOT_EQUALS, Literal: literal}
		} else {
			tok = l.newToken(token.BANG)
		}
	case '*':
		tok = l.newToken(token.ASTERISK)
	case '/':
		tok = l.newToken(token.SLASH)
	case '<':
		tok = l.newToken(token.LESS_THAN)
	case '>':
		tok = l.newToken(token.GREATER_THAN)
	case ',':
		tok = l.newToken(token.COMMA)
	case '(':
		tok = l.newToken(token.PAREN_OPEN)
	case ')':
		tok = l.newToken(token.PAREN_CLOSE)
	case '{':
		tok = l.newToken(token.BRACE_OPEN)
	case '}':
		tok = l.newToken(token.BRACE_CLOSE)
	case '[':
		tok = l.newToken(token.BRACKET_OPEN)
	case ']':
		tok = l.newToken(token.BRACKET_CLOSE)
	case ':':
		tok = l.newToken(token.COLON)
	case ';':
		tok = l.newToken(token.SEMICOLON)
	case '"':
		tok.Pos = l.pos
		tok.Literal = l.readString()
		tok.Type = token.STRING
		tok.Len = len(tok.Literal)
	case '\n':
		l.pos.Ln++
		l.pos.Col = 0
	case 0:
		tok = l.newToken(token.EOF)
	default:
		if isLetter(l.ch) {
			tok.Pos = l.pos
			tok.Literal = l.readIdentifier()
			tok.Type = token.LookupIdentifier(tok.Literal)
			tok.Len = len(tok.Literal)
			return tok
		} else if isDigit(l.ch) {
			tok.Pos = l.pos
			tok.Type = token.INTEGER
			tok.Literal = l.readNumber()
			tok.Len = len(tok.Literal)
			return tok
		} else {
			tok = l.newToken(token.ILLEGAL)
		}
	}

	l.readChar()
	return tok
}

func (l *Lexer) newToken(tt token.TokenType) token.Token {
	return token.Token{Type: tt, Literal: string(l.ch), Pos: l.pos, Len: 1}
}

func (l *Lexer) skipWhitespace() {
	for l.ch == ' ' || l.ch == '\t' || l.ch == '\n' || l.ch == '\r' {
		l.readChar()
	}
}

func (l *Lexer) peekChar() byte {
	if l.readPosition >= len(l.input) {
		return 0
	} else {
		return l.input[l.readPosition]
	}
}

func isLetter(ch byte) bool {
	isLowercaseLetter := 'a' <= ch && ch <= 'z'
	isUppercaseLetter := 'A' <= ch && ch <= 'Z'
	isUnderscore := ch == '_'
	return isLowercaseLetter || isUppercaseLetter || isUnderscore
}

func (l *Lexer) readIdentifier() string {
	position := l.position
	for isLetter(l.ch) {
		l.readChar()
	}
	return l.input[position:l.position]
}

func isDigit(ch byte) bool {
	return '0' <= ch && ch <= '9'
}

func (l *Lexer) readNumber() string {
	position := l.position
	for isDigit(l.ch) {
		l.readChar()
	}
	return l.input[position:l.position]
}

func (l *Lexer) readString() string {
	b := strings.Builder{}

	for {
		l.readChar()
		if l.ch == '\\' {
			switch l.peekChar() {
			case 'n':
				b.WriteByte('\n')
			case 'r':
				b.WriteByte('\r')
			case 't':
				b.WriteByte('\t')
			case '\\':
				b.WriteByte('\\')
			case '"':
				b.WriteByte('"')
			}

			l.readChar()
			continue
		}

		if l.ch == '"' || l.ch == 0 {
			break
		}

		b.WriteByte(l.ch)
	}
	return b.String()
}

File: src/lexer/lexer.zig
const std = @import("std");
const token = @import("../token/token.zig");

pub const Lexer = struct {
    input: []const u8,

    offset: u32 = 0,
    read_offset: u32 = 0,
    ch: u8 = 0,
    pos: token.Pos = .{ .ln = 1, .col = 0 },

    finished: bool = false,

    allocator: std.mem.Allocator,

    pub fn init(allocator: std.mem.Allocator, input: []const u8) Lexer {
        var lexer = Lexer{
            .input = input,
            .allocator = allocator,
        };
        lexer.readChar();
        return lexer;
    }

    pub fn reset(self: *Lexer) void {
        self.offset = 0;
        self.read_offset = 0;
        self.ch = 0;
        self.pos = .{ .ln = 1, .col = 0 };
        self.readChar();
    }

    pub fn next(self: *Lexer) ?token.Token {
        const tok = self.nextToken();

        if (self.finished) {
            return null;
        }

        if (tok.type == .eof) {
            self.finished = true;
        }

        return tok;
    }

    pub fn nextToken(self: *Lexer) token.Token {
        self.skipWhitespace();

        const tok: token.Token = switch (self.ch) {
            // operators
            '=' => blk: {
                if (self.peekChar() == '=') {
                    const start_offset = self.offset;
                    const pos = self.pos;
                    self.readChar();
                    self.readChar();
                    break :blk .{
                        .type = .equals,
                        .literal = self.input[start_offset..self.offset],
                        .pos = pos,
                    };
                } else {
                    break :blk self.newToken(.assign);
                }
            },
            '+' => self.newToken(.plus),
            '-' => self.newToken(.minus),
            '!' => blk: {
                if (self.peekChar() == '=') {
                    const start_offset = self.offset;
                    const pos = self.pos;
                    self.readChar();
                    self.readChar();
                    break :blk .{
                        .type = .not_equals,
                        .literal = self.input[start_offset..self.offset],
                        .pos = pos,
                    };
                } else {
                    break :blk self.newToken(.bang);
                }
            },
            '*' => self.newToken(.asterisk),
            '/' => self.newToken(.slash),
            '<' => self.newToken(.less_than),
            '>' => self.newToken(.greater_than),

            // delitimers
            ',' => self.newToken(.comma),
            ':' => self.newToken(.colon),
            ';' => self.newToken(.semicolon),
            '(' => self.newToken(.paren_open),
            ')' => self.newToken(.paren_close),
            '{' => self.newToken(.brace_open),
            '}' => self.newToken(.brace_close),
            '[' => self.newToken(.bracket_open),
            ']' => self.newToken(.bracket_close),

            // identifiers and literals
            '"' => blk: {
                const pos = self.pos;
                const literal = self.readString() catch "OUT OF MEMORY!!!";
                break :blk .{
                    .type = .string,
                    .literal = literal,
                    .pos = pos,
                };
            },
            0 => self.newToken(.eof),
            else => blk: {
                const pos = self.pos;
                if (std.ascii.isAlphabetic(self.ch)) {
                    const literal = self.readIdentifier();
                    break :blk .{
                        .type = token.lookupIdentifier(literal),
                        .literal = literal,
                        .pos = pos,
                    };
                } else if (std.ascii.isDigit(self.ch)) {
                    break :blk .{
                        .type = .integer,
                        .literal = self.readNumber(),
                        .pos = pos,
                    };
                } else {
                    break :blk self.newToken(.illegal);
                }
            },
        };

        return tok;
    }

    fn newToken(self: *Lexer, tok_type: token.TokenType) token.Token {
        defer self.readChar();
        return .{
            .type = tok_type,
            .literal = self.input[self.offset..self.read_offset],
            .pos = self.pos,
        };
    }

    fn readChar(self: *Lexer) void {
        self.offset = self.read_offset;
        if (self.read_offset >= self.input.len) {
            self.ch = 0;
        } else {
            self.ch = self.input[self.read_offset];
            self.read_offset += 1;
        }

        if (self.ch == '\n') {
            self.pos.ln += 1;
            self.pos.col = 0;
        } else {
            self.pos.col += 1;
        }
    }

    fn skipWhitespace(self: *Lexer) void {
        while (std.ascii.isWhitespace(self.ch)) {
            self.readChar();
        }
    }

    fn peekChar(self: *Lexer) u8 {
        if (self.read_offset >= self.input.len) {
            return 0;
        } else {
            return self.input[self.read_offset];
        }
    }

    fn readString(self: *Lexer) ![]u8 {
        defer self.readChar();
        var literal = std.ArrayList(u8).init(self.allocator);

        while (true) {
            self.readChar();

            if (self.ch == '\\') {
                const escape: u8 = switch (self.peekChar()) {
                    'n' => '\n',
                    'r' => '\r',
                    't' => '\t',
                    '\\' => '\\',
                    '"' => '"',
                    else => 0,
                };

                if (escape != 0) {
                    try literal.append(escape);
                } else {
                    try literal.append(self.ch);
                    try literal.append(self.peekChar());
                }

                self.readChar();
                continue;
            }

            if (self.ch == '"' or self.ch == 0) {
                break;
            }

            try literal.append(self.ch);
        }

        return literal.items;
    }

    fn readIdentifier(self: *Lexer) []const u8 {
        const start_offset = self.offset;
        while (std.ascii.isAlphabetic(self.ch)) {
            self.readChar();
        }
        return self.input[start_offset..self.offset];
    }

    fn readNumber(self: *Lexer) []const u8 {
        const start_offset = self.offset;
        while (std.ascii.isDigit(self.ch)) {
            self.readChar();
        }
        return self.input[start_offset..self.offset];
    }
};

const lexer_test = @import("./lexer_test.zig");

test {
    lexer_test.hack();
}
